treasury.go

```diff
@@ -0,0 +1,204 @@
+	// The first error will be returned to the caller.
+	var firstErr error
+	for _, tHash := range ticketHashes {
+		vspTicketInfo, err := wallet.Internal().VSPTicketInfo(ctx, tHash)
+		if err != nil {
+			// Ignore NotExist error, just means the ticket is not
+			// registered with a VSP, nothing more to do here.
+			if firstErr == nil && !errors.Is(err, errors.NotExist) {
+				firstErr = err
+			}
+			continue // try next tHash
+		}
+
+		// Update the vote policy for the ticket with the associated VSP.
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil && firstErr == nil {
+			firstErr = err
+			continue // try next tHash
+		}
+		err = vspClient.SetVoteChoice(ctx, tHash, nil, nil, policyMap)
+		if err != nil && firstErr == nil {
+			firstErr = err
+			continue // try next tHash
+		}
+	}
+
+	vspPreferenceUpdateSuccess = firstErr == nil
+	return firstErr
+}
+
+// TreasuryPolicies returns saved voting policies for treasury spends per pi key.
+// If a pi key is specified, the policy for that pi key is returned; otherwise the policies
+// for all pi keys are returned. If a ticket hash is provided, the policy(ies) for that ticket
+// is/are returned.
+func (wallet *Wallet) TreasuryPolicies(PiKey, tixHash string) ([]*TreasuryKeyPolicy, error) {
+	var ticketHash *chainhash.Hash
+	if tixHash != "" {
+		tixHash, err := chainhash.NewHashFromStr(tixHash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = tixHash
+	}
+
+	if PiKey != "" {
+		pikey, err := hex.DecodeString(PiKey)
+		if err != nil {
+			return nil, fmt.Errorf("parameter contains invalid hexadecimal: %w", err)
+		}
+		var policy string
+		switch wallet.Internal().TreasuryKeyPolicy(pikey, ticketHash) {
+		case stake.TreasuryVoteYes:
+			policy = "yes"
+		case stake.TreasuryVoteNo:
+			policy = "no"
+		default:
+			policy = "abstain"
+		}
+		res := []*TreasuryKeyPolicy{
+			{
+				PiKey:  PiKey,
+				Policy: policy,
+			},
+		}
+		if tixHash != "" {
+			res[0].TicketHash = tixHash
+		}
+		return res, nil
+	}
+
+	policies := wallet.Internal().TreasuryKeyPolicies()
+	res := make([]*TreasuryKeyPolicy, 0, len(policies))
+	for i := range policies {
+		var policy string
+		switch policies[i].Policy {
+		case stake.TreasuryVoteYes:
+			policy = "yes"
+		case stake.TreasuryVoteNo:
+			policy = "no"
+		}
+		r := &TreasuryKeyPolicy{
+			PiKey:  hex.EncodeToString(policies[i].PiKey),
+			Policy: policy,
+		}
+		if policies[i].Ticket != nil {
+			r.TicketHash = policies[i].Ticket.String()
+		}
+		res = append(res, r)
+	}
+	return res, nil
+}
+
+// PiKeys returns the sanctioned Politeia keys for the current network.
+func (mw *MultiWallet) PiKeys() [][]byte {
+	return mw.chainParams.PiKeys
+}
+
+// PiKey returns a single politiea key (converted to string) from the
+// available politiea keys specified by the provided index.
+func (mw *MultiWallet) PiKey(index int) string {
```

Comment:
This seems buggy, how would callers know which index to pass? Seems like a help function that should be added to godcr perhaps.

Reply:
knowing which index to call is a valid concern, maybe we can add the links to the available keys as part of the comment?

this method seems like it might be more useful in the library

thoughts?

Reply:
There's a `mw` method that returns all the keys, the caller of this `mw.PiKey(index)` method should use that method instead and do...
```
key := hex.EncodeToString(mw.PiKeys()[index])
```